mockrxandroidble/src/main/java/com/polidea/rxandroidble2/mockrxandroidble/RxBleClientMock.java

```diff
@@ -363,7 +389,126 @@ private static boolean filterDevice(RxBleDevice rxBleDevice, @Nullable UUID[] fi
 
     @Override
     public Observable<ScanResult> scanBleDevices(ScanSettings scanSettings, ScanFilter... scanFilters) {
-        return Observable.error(new RuntimeException("not implemented")); // TODO [DS]
+        return createScanOperation(scanSettings, scanFilters);
+    }
+
+    private static ScanResult convertToPublicScanResult(RxBleDevice bleDevice, Integer rssi, ScanRecord scanRecord) {
+        return new ScanResult(bleDevice, rssi, System.currentTimeMillis()*1000000, ScanCallbackType.CALLBACK_TYPE_FIRST_MATCH, scanRecord);
+    }
+
+    @NonNull
+    private ScanResult createScanResult(RxBleDeviceMock rxBleDeviceMock) {
+        return convertToPublicScanResult(rxBleDeviceMock, rxBleDeviceMock.getRssi(), rxBleDeviceMock.getScanRecord());
+    }
+
+    @NonNull
+    private Observable<ScanResult> createScanOperation(ScanSettings scanSettings, final ScanFilter... scanFilters) {
+        return discoveredDevicesSubject
+                .filter(new Predicate<RxBleDeviceMock>() {
+                    @Override
+                    public boolean test(RxBleDeviceMock rxBleDevice) {
+                        return RxBleClientMock.filterDevice(rxBleDevice, scanFilters);
+                    }
+                })
+                .map(new Function<RxBleDeviceMock, ScanResult>() {
+                    @Override
+                    public ScanResult apply(RxBleDeviceMock rxBleDeviceMock) {
+                        return RxBleClientMock.this.createScanResult(rxBleDeviceMock);
+                    }
+                });
+    }
+
+    private static boolean maskedDataEquals(@NonNull byte[] data1, @NonNull byte[] data2, @Nullable byte[] mask) {
+        if(mask == null) {
+            return Arrays.equals(data1, data2);
+        } else {
+            if(data1.length != data2.length || data1.length != mask.length) {
+                return false;
+            }
+            for(int i = 0; i < data1.length; i++) {
+                if((data1[i] & mask[i]) != (data2[i] & mask[i])) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    private static byte[] getDataFromUUID(@NonNull UUID uuid) {
+        ByteBuffer bb = ByteBuffer.wrap(new byte[16]);
+        bb.putLong(uuid.getMostSignificantBits());
+        bb.putLong(uuid.getLeastSignificantBits());
+        return bb.array();
+    }
+
+    private static boolean filterDevice(RxBleDevice rxBleDevice, ScanFilter... scanFilters) {
```

Comment:
I am not sure why it is `RxBleDevice` here where it is clearly meant for `RxBleDeviceMock`?

Reply:
Me neither...copied from the original method `private static boolean filterDevice(RxBleDevice rxBleDevice, @Nullable UUID[] filterServiceUUIDs) {`

Reply:
Do I need to change this? I don't think it matters really.